packages/sources/alongside/src/endpoint/collateral/utils.ts

```diff
@@ -0,0 +1,249 @@
+      lastFeeRate = feeRate
+      lastTimestamp = timestamp
+    }
+    return inflation
+  }
+
+  calcInflation = (feeRate: number, fromTimestamp: number, toTimestamp: number) => {
+    const bigNumResult = bignumber(1)
+      .div(bignumber(1).plus(feeRate))
+      .pow(
+        bignumber(toTimestamp)
+          .sub(bignumber(fromTimestamp))
+          .div(60 * 60 * 24),
+      )
+    return number(bigNumResult)
+  }
+
+  getLatestMethodologySetInfo = async (indexToken: ethers.Contract) => {
+    const latestBlock = await this.provider.getBlock('latest')
+    const logs = await this.provider.getLogs({
+      ...indexToken.filters.MethodologySet(),
+      fromBlock: latestBlock.number - 1_000_000,
+      toBlock: latestBlock.number,
+    })
+    // TODO replace me with an assert
+    if (logs.length === 0) {
+      throw new Error("shouldn't happen")
+    }
+    const blockNumber = logs[logs.length - 1].blockNumber
+    return await this.withTimestamp(blockNumber)
+  }
+
+  getFeeChanges = async (indexToken: ethers.Contract, fromBlock: number, toBlock: number) => {
+    return await Promise.all(
+      (
+        await this.provider.getLogs({
+          ...indexToken.filters.FeeRateSet(),
+          // TODO test that this won't mess up when they're the same number
+          // TODO test we exclude the fee change when
+          // it's _before_ the methodology set in the same block
+          fromBlock,
+          toBlock,
+        })
+      ).map(async (feeChange) => [
+        await this.getTimestamp(feeChange.blockNumber),
+        parseInt(feeChange.topics[1]) / Math.pow(10, 20),
+      ]),
+    )
+  }
+
+  getTimestamp = async (blockNumber: number) =>
+    (await this.provider.getBlock(blockNumber)).timestamp
+
+  withTimestamp = async (blockNumber: number) => ({
+    blockNumber,
+    timestamp: await this.getTimestamp(blockNumber),
+  })
+
+  fetchMethodology = async () =>
+    (
+      await this.fetchIPFS(
+        await this.IndexToken.methodology({
+          blockTag: (await this.provider.getBlock('latest')).number,
+        }),
+        false,
+      )
+    ).assets
+
+  fetchIPFS = async (path: string, pinataFirst = true) => {
+    try {
+      if (pinataFirst) {
+        const data = await this.fetchIPFSFromPinata(path)
+        return data
+      } else {
+        const data = await this.fetchIPFSFromInfura(path)
+        return data
+      }
+    } catch (error) {
+      console.log(
+        `failed to fetch from ${pinataFirst ? 'Pinata' : 'Infura'}, trying ${
+          pinataFirst ? 'Infura' : 'Pinata'
+        }"}`,
+      )
+      if (pinataFirst) {
+        const data = await this.fetchIPFSFromInfura(path)
+        return data
+      } else {
+        const data = await this.fetchIPFSFromPinata(path)
+        return data
+      }
+    }
+  }
+
+  calcMinCollateral = (
+    tradingBalances: BalanceType[],
+    vaultBalances: BalanceType[],
+    units: { [x: string]: number },
+  ): number => {
+    console.log({
```

Comment:
We should move this to using a logger and any other place I've missed that has console.log